lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs

```diff
@@ -334,9 +335,14 @@ instance CompareDiscovery (RndState n) where
 
 instance KnownAddresses (RndState n) where
     knownAddresses s = mconcat
-        [ Map.elems (discoveredAddresses s)
-        , Map.elems ((,Unused) <$> pendingAddresses s)
+        [ retrieveAddrsWithPaPath (discoveredAddresses s)
+        , retrieveAddrsWithPaPath ((,Unused) <$> pendingAddresses s)
         ]
+      where
+        constructAddrWithPath path (addr,state) acc =
+            (addr, state, toDerivationIndexes path):acc
+        retrieveAddrsWithPaPath =
+            Map.foldrWithKey constructAddrWithPath []
```

Comment:
Suggestion to avoid the extra traversal on pending addresses and also, make the function a bit more symmetric to more clearly see how the output is constructed:

```hs
instance KnownAddresses (RndState n) where
    knownAddresses s = mconcat
        [ toListWithPath (\path (addr, state) -> (addr, state, path))
            (discoveredAddresses s)
        , toListWithPath (\path addr -> (addr, Unused, path))
            (pendingAddresses s)
        ]
      where
        toListWithPath
            :: (NonEmpty DerivationIndex -> v -> result)
            -> Map DerivationPath v
            -> [result]
        toListWithPath mk =
            Map.foldrWithKey
                (\path v result -> mk (toDerivationIndexes path) v : result)
                []
```